### PR TITLE
Implemented inline assembly

### PR DESCRIPTION
**Pull request by tillspeicher**

### CODE REVIEW

File: src/impala/emit.cpp [hunk 1]
```diff
@@ -731,9 +731,10 @@ void AsmStmt::emit(CodeGen& cg) const {
     for (auto expr : input_exprs_)
         inputs[i++] = cg.remit(expr);
 
-    const Asm *asm_op = cg.world().inl_asm(cg.get_mem(), out_types, inputs, loc(),
+    const Assembly *asm_op = cg.world().assembly(out_types, cg.get_mem(), inputs,
```
Comment (leissa) [inline src/impala/emit.cpp]: use `auto assembly = cg.world().assembly(...)`

**Comment by tillspeicher** [comment]
It should be ready to be merged now.

**Comment by leissa** [comment]
I'd like to get rid of the macro/goto magic in the parser. Can we use `parse_comma_list` instead?

**Comment by tillspeicher** [comment]
We cannot use `parse_comma_list` to replace `parse_asm_operands` because it only checks for one delimiter to stop parsing, but usually there can be `COLON`, `DOUBLE_COLON` and `R_PAREN` as delimiters.
However, we can use it to get rid of the second call to `get_asm_options`, then we can eliminate that function.

Getting rid of the marco/gotos is not so easy because of the double colons.
With just colons something like nested ifs would work but with double colons you may have to jump completely over parsing an operand list.
Two solutions without `goto`s would be to 1) have `if`s + flags to jump over parsing operand lists or 2) to have an array of functions which each parse one operand list and let the colons and double colons advance an index into that array.
I think the second option would be nice but it would break with the style of the code in the rest of the parser (as do the `goto`s).

File: src/impala/emit.cpp [hunk 2]
```diff
@@ -728,7 +728,7 @@ void AsmStmt::emit(CodeGen& cg) const {
     for (size_t i = 0, e = num_inputs(); i != e; ++i)
         ins[i] = cg.remit(input(i).expr());
 
-    thorin::Assembly::Flags flags;
+    thorin::Assembly::Flags flags = thorin::Assembly::Flags::NoFlag;
```
Comment (leissa) [inline src/impala/emit.cpp]: thanks for spotting this bug :)

File: src/impala/parser.cpp [hunk 3]
```diff
@@ -1294,8 +1294,6 @@ const AsmStmt* Parser::parse_asm_stmt() {
     return asm_stmt;
 }
 
-#include <iostream>
-
```
Comment (leissa) [inline src/impala/parser.cpp]: Can you plz remove this...

Comment (tillspeicher) [inline src/impala/parser.cpp]: What do you mean? I removed it in this commit, the other one too.

Comment (leissa) [inline src/impala/parser.cpp]: oh yes, my bad. just too stupid to read a diff...

**Pull request merged**